### ISSUE TITLE
Additional F-Droid repository for nightly

### ISSUE THREAD

**Issue by JensKorte** [kiwix/kiwix-android#1777]
**Is your feature request related to a problem? Please describe.**
I have no Playstore Account and therefore I cannot get the nightly version easily. I have to download it from <https://download.kiwix.org/nightly/>.

**Describe the solution you'd like**

It would be great, if kiwix had a F-Droid repository for nightlies like repos listed in <https://forum.f-droid.org/t/known-repositories/721>. See also: <https://f-droid.org/en/repomaker/>

**Describe alternatives you've considered**
F-Droid themselves could have a nightly/developer repository where the beta/nightly versions of apps are listed.

**Additional context**
Since this is a very low priority issue this issue can be closed quite fast.

**Comment by kelson42** [comment]
Sounds doable...

**Comment by dipu989** [comment]
@kelson42 @macgills Would like to do this! Please assign.

**Comment by dipu989** [comment]
@kelson42 @macgills We need to make a Kiwix repo on Fdroid that hosts our nightly builds via Gitlab? Is the approach correct?
References - https://android.izzysoft.de/articles/named/fdroid-intro-3?lang=en

**Comment by macgills** [comment]
[This](https://gitlab.com/fdroid/fdroiddata/-/blob/4f3c79e40009fb930ec0ff9d4415a26fe2264212/metadata/org.kiwix.kiwixmobile.yml) is the file we have on fdroids repository that dictates what gets hosted on fdroid.

I really can't help you much as I have only ever done 1 update to that file and most of fdroid is still very confusing to me. Maybe the repomaker link in the description of this ticket is useful?

**Comment by JensKorte** [comment]
The repomaker is imho the best way. You can create your own repo on your own server and a user can add the repo adress (https://...) in F-Droid at (translated from the german F-Droid) "Options" -> "packet sources" -> "+"

**Comment by kelson42** [comment]
@JensKorte @macgills Please don't mix with #1777. #1777 is not the solution for this ticket. I want Kiwix is the standard FDroid repo and this is the most important. The solution here is just to make the FDroid PR work fine (or does the option to avoid signing just fail like suggested in FDroid PR?).

**Comment by macgills** [comment]
@kelson42 this is #1777 .

I was just giving dipu the only information I have as regards fdroid. I don't know how automatic publishing will work because we certainly can't set it up for the actual releases of the app on fdroid so I would think this ticket is mildly doomed for the time being

**Comment by kelson42** [comment]
@macgills Sorry... In addition to push to download.kiwix.org we should push to this repo. (wherever he is...). So this is not about releases, this is only for nightlies.

**Comment by macgills** [comment]
The new repo doesn't work like how the fdroid repo does? We can just hit an endpoint with an apk? If so perfect, if we can't and have to do it the way fdroid do it we are stuck the same way we are for our real releases

**Comment by JensKorte** [comment]
> The new repo doesn't work like how the fdroid repo does? We can just hit an endpoint with an apk? If so perfect, if we can't and have to do it the way fdroid do it we are stuck the same way we are for our real releases

I am not sure, if I get you right. In an F-Droid repo there can be different versions of a program, there is a recommended version and the user can select to up-/downgrade to an alpha or older version.
kiwix-nightly could be the package name, then it is not mixed up with kiwix. If you want to provide test versions like different branches, I would suggest to provide an extra package with a name like kiwix-<branchname>.

**Comment by macgills** [comment]
Yes but I currently don't believe we have an easy way of automatically updating this "nightly repo" as I would assume the update mechanism is similar to on the main fdroid repo which we currently cannot support because we need to supply an endpoint with version codes.

Our nightly builds are also development versions if i remember correctly, they are intended for nobody but developers.

**Comment by JensKorte** [comment]
> Our nightly builds are also development versions if i remember correctly, they are intended for nobody but developers.

I am no developer, but I can help with beta testing.

**Comment by gouri-panda** [comment]
@kelson42 @MohitMaliFtechiz Since we are on F-Droid now. Now it's the perfect time to deploy nightly automatically. I can make a PR that CI automatically puts nightly APK to F-Droid.

**Comment by kelson42** [comment]
@MohitMaliDeveloper  @gouri-panda It seems this has not been fixed fully. Please explain what is the current problem.

**Comment by gouri-panda** [comment]
@kelson42 According to documentation We have to follow these steps:
 * Create a new GitLab “project” by appending -nightly to the name. For example, https://gitlab.com/fdroid/fdroidclient becomes https://gitlab.com/fdroid/fdroidclient-nightly and https://gitlab.com/eighthave/fdroidclient becomes https://gitlab.com/eighthave/fdroidclient-nightly

* In that new project, add the SSH Public Key derived from your debug.keystore as a Deploy Key in your Repository Settings, e.g. https://gitlab.com/eighthave/fdroidclient-nightly/settings/repository

* In the CI/CD Settings of the project being built, paste that into GitLab CI Secret Variable called DEBUG_KEYSTORE, e.g.https://gitlab.com/eighthave/fdroidclient/settings/ci_cd

**Comment by kelson42** [comment]
@gouri-panda If i create the repo, you can do the rest?

**Comment by gouri-panda** [comment]
@kelson42 Yes!

**Comment by kelson42** [comment]
https://github.com/kiwix/kiwix-android-nightly

**Comment by kelson42** [comment]
@MohitMaliFtechiz Time to relaunch this. We should first decide if we should do it. Do you think this is doable in good conditions and that we could achieve something like we do with PlayStore weekly dev version... but on FDroid?

**Comment by kelson42** [comment]
@MohitMaliFtechiz Regarding F-Droid re-integration, I don't  think this is a good idea to go with release `3.14.0`. Reason is that this version (1) has still ObjectBox code (2) is in many ways already old.

I would recommend to achieve to push our nightlies to FDroid first, and then once this is done, work to prepare the `4.0.0` publishing on FDroid.

Does that sounds a good plan to you?

**Comment by MohitMaliFtechiz** [comment]
@kelson42 Sounds good to me.

**Comment by MohitMaliFtechiz** [comment]
@kelson42 Does this repo still exist https://github.com/kiwix/kiwix-android-nightly? I can not see it.

**Comment by MohitMaliFtechiz** [comment]
@licaon-kter, @eighthave, can you please tell us why we need the https://github.com/kiwix/kiwix-android-nightly repository? As mentioned in the docs https://f-droid.org/docs/Publishing_Nightly_Builds/. Looking at the example shown in the documentation, it seems this nightly repository is for storing APKs and metadata: https://github.com/zom/Zom-Android-nightly. For this, we need to create a nightly repo? If yes, can we store these things outside of GitHub? Like on our server(Then how can we configure this?). Or is the GitHub repo for other things?

**Comment by kelson42** [comment]
@licaon-kter Like @MohitMaliFtechiz said,  I would like to consider (1) building nightly on kiwix/kiwix-androis but (2) host (FDroid nightly repo) outside, ideally would be a container image providing this feature.

**Comment by licaon-kter** [comment]
sure, use the docker, use a local cloned fdroidserver, as you can fit into your workflow :)

**Comment by MohitMaliFtechiz** [comment]
@licaon-kter Thanks for your quick response. Is there any guide or doc to do it?

**Comment by licaon-kter** [comment]
I personally don't host so not sure how the https://f-droid.org/docs cover this, it's just a matter of `fdroid update` and `fdroid publish`   and host a folder on the net :)

**Comment by MohitMaliFtechiz** [comment]
@licaon-kter One last question in this, we only need to publish the APK at a place on the internet. So, we are already publishing the nightly APK(but with a release signing key to make it easily updatable), so could we configure the already published APK with fdroid nightly? See all the APKs are published here in folders https://download.kiwix.org/nightly/. And APK list like this https://download.kiwix.org/nightly/2025-08-29/. It will work with fdroid? or it need a dedicated folder?

**Comment by licaon-kter** [comment]
@MohitMaliFtechiz yes, eg. https://ftp.agdsn.de/fdroid/ see repo/ folder

**Comment by MohitMaliFtechiz** [comment]
@licaon-kter Okay, thank you. @kelson42, Can you please create a folder named `fdroid` on https://download.kiwix.org/?

**Comment by licaon-kter** [comment]
@MohitMaliFtechiz actual contents are in fdroid/repo/ fyi, I can't link for that example since you'll see our special index :)


normally you'll see the onboarding page with QR for users to scan from F-Droid client to add the repo

**Comment by kelson42** [comment]
@MohitMaliFtechiz Please explain first from both user, infrastructure and CD worflow perspective what is your plan to implement this issue.

**Comment by MohitMaliFtechiz** [comment]
@kelson42 I have researched about how this can be done via our server, and found these things:

First, we need to create https://download.kiwix.org/fdroid/repo. From where fdroid fetches our nightly APKs. Inside this folder, we will add the `icon.png` inside the `icons` folder. That is our Kiwix app icon. We will upload it via CI.
## In our CI ##
* We will install the fdroid server. After generating the APK.
* Then we will copy the generated apk and kiwix icon inside the temp `fdroid/repo`.
* Inside that folder, we will run `fdroid update --create-metadata` // This generates an index-v1.jar and index.xml, which F-Droid clients use to list apps.
* Then, we upload this to our server inside the `fdroid/repo` directory.

For users: It will be hosted like this https://fdroid.storchp.de/fdroid/repo/.

@licaon-kter Does it sound good to you?

**Comment by kelson42** [comment]
@MohitMaliFtechiz user perspective is missing? How do user know about our new nightlies and how do they upgrade?

**Comment by MohitMaliFtechiz** [comment]
@kelson42 The user can add our nightly repo in their fdroid app, and can update the application whenever a new update is published.

In the video below, I have taken the  https://fdroid.storchp.de/fdroid/repo/ for example.
In the fdroid app:
* User goes to settings -> Repositories -> Click on plus icon, and scan the QR to add our nightly repo in thier fdroid app. (We will add this QR and about it in the readme file once our nightly repo is published on fdroid).
* After adding the repository, they can click on it, install or update the app(Maybe fdroid will send a notification when a new update is available).

https://github.com/user-attachments/assets/8caffafd-bc8d-45e9-9e91-0c24659528ae

**Comment by kelson42** [comment]
I finally come back on this issue, sorry for the long delay. Reasons is that I'm (still) unsure about the way to go:

For the moment we have such distribution channels:
* Google Play (a version with a few limitations due to Google policy): releases, custom beta channel, custom internal channel (weekly push)
* F-Droid (full version): releases
* download.kiwix.org: releases and nightlies

Now we:
* Discuss to launch another custom F-Droid custom repository to publish nightlies.
* We have a work ongoing to [allow people downloading releases from download.kiwix.org to be automatically informed about updates](https://github.com/kiwix/kiwix-android/issues/4215).

I kind of think it would be good to wrap-up and be sure we want all of this old + new stuff to happen. I see in particular some kind of duplicate things between a Kiwix F-Droid (compatible) custom repository and what is provided anyway on download.kiwix.org (nightlies and releases).

@rgaudin @MohitMaliFtechiz What do you think?

@MohitMaliFtechiz Two questions for you:
* Concerning a Kiwix F-Droid repository, do you need only an access to upload files to https://download.kiwix.org/fdroid/repo? Or do we need something more (who is in charge to provide [the full index or the HTML pages](https://fdroid.storchp.de/fdroid/repo/))?
* Would one F-Droid repository allows to have both "tracks": nightlies and releases?

**Comment by mnalis** [comment]
> Discuss to launch another custom F-Droid custom repository to publish nightlies.

My understanding is that "F-droid custom repo" was intended primarily as a workaround for F-droid users back in a day, since official F-droid themselves wouldn't publish "binary-blob" versions of Kiwix, so there is no way for F-droid users to get the app (release _or_ nightly)

> F-Droid (full version): releases

Which Kiwix still (to the best of my knowledge) does **not** have? 
(but it should become available "soon", as blocking issues have been mostly resolved IIUC?)

Once a regular F-droid Kiwix releases become a thing (hopefully soon), I personally do not see much use of custom F-droid nightly builds. Just making a regular releases periodic enough (_["release early, release often"](https://en.wikipedia.org/wiki/Release_early%2C_release_often)_ mantra) and it should be good enough I think.

**Comment by MohitMaliFtechiz** [comment]
> Concerning a Kiwix F-Droid repository, do you need only an access to upload files to https://download.kiwix.org/fdroid/repo? Or do we need something more (who is in charge to provide [the full index or the HTML pages](https://fdroid.storchp.de/fdroid/repo/))?

Where I was left implementing this in https://github.com/kiwix/kiwix-android/pull/4437, there was an error in uploading the generated metadata on download.kiwix.org/fdroid/repo. After uploading this, we need to inform the fdroid client(via fdroid publish) so that it can publish the update. I am not sure right now, we need to do more as after uploading we will see if the `fdroid publish` will give an error or not. But for the HTML pages and other stuffs, it is automatically generated by the `fdroid update --create-metadata` command.

> Would one F-Droid repository allows to have both "tracks": nightlies and releases?

No, both can not be installed since both have the same package names. So the user can only install a track at a time. Regarding the tracks allowed by the F-Droid repository. Yes, it is supported. We have the main release track, which is currently paused due to objectbox(but will activate this), and the nightly track will be fetched from the download.kiwix.org/fdroid/repo, fdroid suggests creating a new repo for nightly, appending the -nightly at the end. But we are publishing it on our server, so it will be different from our main repository.

**Comment by MohitMaliFtechiz** [comment]
@kelson42 We can move this one and https://github.com/kiwix/kiwix-android/issues/3083 to the next milestone. Since there are still some efforts/testing left, and this is for publishing the nightly on fdroid, this PR code is not mandatory for release. It can happen after release.

**Comment by kelson42** [comment]
I'm less and less convinced we will build a full repository compatible with FDroid to put nightly. I'm almost reqdy to close this issue as WONTFIX, but just postponing.
